pkg/providers/vsphere/virtualmachine/cdrom.go

```diff
@@ -71,9 +70,33 @@ func UpdateCdromDeviceChanges(
 		expectedBackingFileNameToCdrom[bFileName] = cdrom
 	}
 
-	// Ensure all CD-ROM devices are assigned to controllers with proper keys.
+	// Remove any existing CD-ROM devices that do not have the expected backing.
+	// Add them to the device changes with "Remove" operation.
+	// Also, update the current device list by excluding the removed CD-ROMs to
+	// assign the new CD-ROMs to the correct controller unit numbers later.
+	newCurDevices := object.VirtualDeviceList{}
+	for _, d := range curDevices {
+		cdrom, ok := d.(*vimtypes.VirtualCdrom)
+		if !ok {
+			newCurDevices = append(newCurDevices, d)
+			continue
+		}
+
+		if b, ok := cdrom.Backing.(*vimtypes.VirtualCdromIsoBackingInfo); ok &&
+			expectedBackingFileNameToCdrom[b.FileName] != nil {
+			newCurDevices = append(newCurDevices, cdrom)
+			continue
+		}
+
+		deviceChanges = append(deviceChanges, &vimtypes.VirtualDeviceConfigSpec{
+			Device:    cdrom,
+			Operation: vimtypes.VirtualDeviceConfigSpecOperationRemove,
+		})
+	}
+
+	// Ensure all CD-ROM devices are assigned to controllers with proper slots.
 	// This may result new controllers to be added if none is available.
-	newControllers := ensureAllCdromsHaveControllers(expectedBackingFileNameToCdrom, curDevices)
+	newControllers := ensureAllCdromsHaveControllers(expectedBackingFileNameToCdrom, newCurDevices)
```

Comment:
Hey @dougm - Am I just blind where`VirtualDeviceList::AssignController()` updates the picked Controller's `Device` list? Otherwise, I think you can over assign to a controller and end up with a -1 UnitNumber.

Reply:
Yeah, we should probably adopt the EnsureDisksHaveControllers pattern of checking if controller is maxed out and creating a new controller if so. https://github.com/vmware/govmomi/blob/main/vim25/types/configspec.go#L254

Reply:
@bryanv Good catch! I tried deploying a VM with 3 CD-ROMs. and they were all assigned to the same IDE controller, failing the VM reconfiguration task since IDE controllers only have 2 slots.

To fix this, I think we could either rewrite the `AssignController` in govmomi to update the controller's device, or do that in our `ensureAllCdromsHaveControllers` function, e.g.:

```diff
--- a/pkg/providers/vsphere/virtualmachine/cdrom.go
+++ b/pkg/providers/vsphere/virtualmachine/cdrom.go
@@ -301,8 +301,11 @@ func ensureAllCdromsHaveControllers(

                if ide := curDevices.PickController((*vimtypes.VirtualIDEController)(nil)); ide != nil {
                        curDevices.AssignController(cdrom, ide)
+                       ide.GetVirtualController().Device = append(ide.GetVirtualController().Device, cdrom.GetVirtualDevice().Key)
                } else if sata := curDevices.PickController((*vimtypes.VirtualSATAController)(nil)); sata != nil {
                        curDevices.AssignController(cdrom, sata)
+                       sata.GetVirtualController().Device = append(sata.GetVirtualController().Device, cdrom.GetVirtualDevice().Key)
```

Reply:
PR to fix this from the govmomi side: https://github.com/vmware/govmomi/pull/3588

Reply:
Resolving this as govmomi version has been updated to consume the above fix. Also added validation result in the PR's description. Thanks again @bryanv and @dougm for caching this!